### PR TITLE
beam-packages: drop erlang R18 R19 R20

### PR DESCRIPTION
**Pull request by DianaOlympos**
###### Motivation for this change
These versions are not supported by upstream anymore

I have not dropped R21 because R24 is still a bit of a pain in some situations

Yaws needed R18, if this breaks yaws, i think we may have to drop it. @cillianderoiste is the maintainer, maybe he want to chim in.

Also i did not put anything int he release notes.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by happysalada** [comment]
Really like the idea!
ofborg is complaning because of this line
`  inherit (beam.packages.erlangR19) cuter lfe_1_2;`

@dlesl @gleber you might be interested in this

**Comment by dlesl** [comment]
I agree with these changes; those who want to use these versions can pin to an old version of nixpkgs (since security shouldn't be an issue anyway) or use an overlay like https://github.com/id3as/nixpkgs-nixerl

**Comment by gleber** [comment]
:+1: to this change from me.

**Comment by DianaOlympos** [comment]
I updated the docs to use R22 as a default and i made cuter use R21 and lfe_1_9 use R21. If they cannot compile on this, i think we will have to mark them as broken.

**Comment by DianaOlympos** [comment]
i dropped lfe_1_2, it is simply not available on erlang > 19

File: doc/languages-frameworks/beam.section.md [hunk 1]
```diff
@@ -8,9 +8,9 @@ In this document and related Nix expressions, we use the term, _BEAM_, to descri
 
 All BEAM-related expressions are available via the top-level `beam` attribute, which includes:
 
-- `interpreters`: a set of compilers running on the BEAM, including multiple Erlang/OTP versions (`beam.interpreters.erlangR19`, etc), Elixir (`beam.interpreters.elixir`) and LFE (Lisp Flavoured Erlang) (`beam.interpreters.lfe`).
+- `interpreters`: a set of compilers running on the BEAM, including multiple Erlang/OTP versions (`beam.interpreters.erlangR22`, etc), Elixir (`beam.interpreters.elixir`) and LFE (Lisp Flavoured Erlang) (`beam.interpreters.lfe`).
 
-- `packages`: a set of package builders (Mix and rebar3), each compiled with a specific Erlang/OTP version, e.g. `beam.packages.erlangR19`.
+- `packages`: a set of package builders (Mix and rebar3), each compiled with a specific Erlang/OTP version, e.g. `beam.packages22`.
```
Comment (happysalada) [inline doc/languages-frameworks/beam.section.md:13]: packages22 -> packages.erlangR22

**Comment by happysalada** [comment]
Result of `nixpkgs-review pr 123333` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>cuter</li>
  </ul>
</details>

**Comment by DianaOlympos** [comment]
do we want to mark it as broken ? i do not want to touch it and i have no idea how many people even use it.

**Comment by happysalada** [comment]
I would say yes, mark it as broken.

@ericbmerritt We would like to drop old erlang versions. Would you be willing to update cuter? or are you okay to mark it as broken?

I would say let's wait a couple of days before merging.

**Comment by DianaOlympos** [comment]
as long as we get this in 21.05. @jonringer all good for you if we wait a couple days to see if we get buyin ? Otherwise we mark cuter as broken and we can fix later

**Review by jonringer — APPROVED** [review]
If they aren't supported by upstream, then these should be removed.

**Comment by jonringer** [comment]
Do you mind adding some release notes saying that they were dropped

**Comment by jonringer** [comment]
seems like cuter is having issues, however, it's marked as insecure on master. So.... yea

```
https://github.com/NixOS/nixpkgs/pull/123333

1 package failed to build:
cuter

1 package built:
yaws
```

**Comment by happysalada** [comment]
@DianaOlympos I would say we waited long enough. If you are ok to deprecate cuter, and add a couple of notes around the deprecation, I'm happy to merge this anytime.

**Comment by DianaOlympos** [comment]
Yeah i will see if i can get this done this weekend. Conference stuff plus work have me swamp until then

**Comment by dlesl** [comment]
Rebar3 0.16.0 (which just came out) drops support for OTP 21. Maybe we should consider doing the same since having multiple versions of rebar3 could get messy.

**Comment by happysalada** [comment]
@DianaOlympos no rush! Thanks a lot for taking the time to do this!

@dlesl I like the idea to drop more versions. However, elixir's deprecation schedule keeps 1.7 around for 6 more months and minimum OTP is 19. Regarding this, I would say, let's keep one version of rebar3 and keep OTP 19 as well. erlang users, would know what version they need to use. As long as the default erlang is the correct version, then a user would have to go out of his way to make a mistake.

**Comment by DianaOlympos** [comment]
I dropped cuter, it was just an example anyway. I added a note in the release notes.

This will need to be backported to the 21.05 right @jonringer ?

**Comment by jonringer** [comment]
> This will need to be backported to the 21.05 right @jonringer ?

If they are meant to be dropped for the release, correct. It appears that you have added the release notes, so I assume that is your intention :).

I'm okay with backporting, if you were looking for approval :+1:

**Comment by DianaOlympos** [comment]
Yeah i would love it to get backported. I am just not used to the way it works during the release window, so i was more checking for making sure i understood the process :D

**Comment by jonringer** [comment]
> Yeah i would love it to get backported. I am just not used to the way it works during the release window, so i was more checking for making sure i understood the process smile

We largely just capture what occurs on master. And try to avoid situations where there will be software which will EOL before the next release.

This seems more like housekeeping, so I think we should move forward with it

**Comment by happysalada** [comment]
@DianaOlympos Thanks a lot for this!

**Comment by DianaOlympos** [comment]
@happysalada do you want to do the port to stable ? i do not have the time this week and we are getting close to cutoff

**Pull request merged**